types/src/v0/impls/stake_table.rs

```diff
@@ -20,27 +22,94 @@ use thiserror::Error;
 use url::Url;
 
 type Epoch = <SeqTypes as NodeType>::Epoch;
-type SeqStakeTableEntry = <SignatureKey as hotshot::types::SignatureKey>::StakeTableEntry;
-type SignatureKey = <SeqTypes as NodeType>::SignatureKey;
+
+// NewTypes for two types of stake tables to avoid confusion
+#[derive(Clone, Debug, From, Into)]
+pub struct DAStakeTable(pub Vec<StakeTableEntry<PubKey>>);
+
+#[derive(Clone, Debug, From, Into)]
+pub struct ConsensusStakeTable(pub Vec<StakeTableEntry<PubKey>>);
+
+#[derive(Clone, Debug)]
+pub struct StakeTables {
+    pub consensus_stake_table: ConsensusStakeTable,
+    pub da_stake_table: DAStakeTable,
+}
+
+impl StakeTables {
+    pub fn new(consensus_stake_table: ConsensusStakeTable, da_stake_table: DAStakeTable) -> Self {
+        Self {
+            consensus_stake_table,
+            da_stake_table,
+        }
+    }
+
+    /// Create the consensus and DA stake tables from L1 events
+    ///
+    /// This is a pure function, to make it easily testable.
+    ///
+    /// We expect have at most a few hundred EVM events in the
+    /// PermissionedStakeTable contract over the liftetime of the contract so it
+    /// should not significantly affect performance to fetch all events and
+    /// perform the computation in this functions once per epoch.
+    pub fn from_l1_events(updates: Vec<StakersUpdatedFilter>) -> Self {
+        let changes_per_node = updates
+            .into_iter()
+            .flat_map(|event| {
+                event
+                    .removed
+                    .into_iter()
+                    .map(|node_info| StakeTableDelta::remove(node_info.into()))
+                    .chain(
+                        event
+                            .added
+                            .into_iter()
+                            .map(|node_info| StakeTableDelta::add(node_info.into())),
+                    )
+            })
+            .group_by(|delta| delta.node_info.stake_table_key);
+
+        // If the last event for a stakers is `Added` the staker is currently
+        // staking, if the last event is removed or (or the staker is not present)
+        // they are not staking.
+        let currently_staking = changes_per_node
```

Comment:
can a node be added and then removed or vice versa?

Reply:
so all the deltas of a node will have removed events first and then added? What happens if it is added and then removed?

Reply:
Yes, the removed events are processed first.

So you can remove a node and in the same transaction add it again (to update the keys or DA status of that node).

Adding and removing is possible but not inside the same Ethereum transaction because the removals are processed first.

Reply:
so `Vec<StakersUpdatedFilter>` is all the events from the start till the current block right? so we may do that in different transactions but always process the removed one first?

Reply:
A transaction sent to the permissioned stake table contract will emit one event which will give us one `StakersUpdatedFilter`. That type has two Vecs of NodeInfo for removing and adding nodes respectively.

When doing the query to ethers one can set the "from block" and the "to block" to fetch a subset of events. Currenly here we would assume that we set "from block" to zero but "to block" we need to set correctly. So that we exclude changes during the current epoch until the next epoch.